askama_parser/src/node.rs

```diff
@@ -37,38 +37,28 @@ pub enum Node<'a> {
 
 impl<'a> Node<'a> {
     pub(super) fn parse_template(
-        i: &mut &'a str,
+        i: &mut InputStream<'a>,
         s: &State<'_, '_>,
     ) -> ParseResult<'a, Vec<Box<Self>>> {
-        let start = *i;
-        let result = match (|i: &mut _| Self::many(i, s)).parse_next(i) {
-            Ok(result) => result,
-            Err(err) => {
-                if let ErrMode::Backtrack(err) | ErrMode::Cut(err) = &err
-                    && err.message.is_none()
-                {
-                    *i = start;
-                    if let Some(mut span) = err.span.as_suffix_of(i) {
-                        opt(|i: &mut _| unexpected_tag(i, s)).parse_next(&mut span)?;
-                    }
-                }
-                return Err(err);
-            }
-        };
```

Comment:
I moved the logic of this part into its own function, because `cut_node` works the same. No need to have the same code twice.